modin/core/dataframe/pandas/partitioning/partition_manager.py

```diff
@@ -641,7 +641,24 @@ def to_pandas(cls, partitions):
         pandas.DataFrame
             A pandas DataFrame
         """
-        retrieved_objects = [[obj.to_pandas() for obj in part] for part in partitions]
+        retrieved_objects = cls.get_objects_from_partitions(partitions.flatten())
+        if all(
+            isinstance(obj, (pandas.DataFrame, pandas.Series))
+            for obj in retrieved_objects
+        ):
+            height, width, *_ = tuple(partitions.shape) + (0,)
+            # restore 2d array
+            retrieved_objects = [
+                [retrieved_objects[i * width + j] for j in range(width)]
+                for i in range(height)
+            ]
```

Comment:
I wonder if something like
```python
objs = iter(retrieved_objects)
retrieved_objects = [[next(objs) for _ in range(width)] for __ in range(height)]
```
actually reads better (it probably works slightly faster because of no index arithmetic)

Reply:
I like your implementation. Rewrited